lib/AST/GenerateAST.cpp

```diff
@@ -136,78 +137,119 @@ static std::string GetName(llvm::Value *v) {
   return s;
 }
 
-z3::expr GenerateAST::GetOrCreateEdgeForBranch(llvm::BranchInst *inst,
+unsigned GenerateAST::GetOrCreateEdgeForBranch(llvm::BranchInst *inst,
                                                bool cond) {
-  if (z_br_edges.find({inst, cond}) == z_br_edges.end()) {
-    if (cond) {
+  auto ToExpr = [&](unsigned idx) {
+    if (idx == poison_idx) {
+      return dec_ctx.z3_ctx.bool_val(false);
+    }
+    return dec_ctx.z3_exprs[idx];
+  };
+  if (dec_ctx.z3_br_edges.find({inst, cond}) == dec_ctx.z3_br_edges.end()) {
+    if (auto constant =
+            llvm::dyn_cast<llvm::ConstantInt>(inst->getCondition())) {
+      dec_ctx.z3_br_edges[{inst, cond}] = dec_ctx.z3_exprs.size();
+      auto edge{dec_ctx.z3_ctx.bool_val(constant->isOne() == cond)};
+      dec_ctx.z3_exprs.push_back(edge);
+      dec_ctx.z3_br_edges_inv[edge.id()] = {inst, true};
+    } else if (cond) {
       auto name{GetName(inst)};
-      auto edge{z_ctx->bool_const(name.c_str())};
-      z_br_edges[{inst, cond}] = z_exprs.size();
-      z_exprs.push_back(edge);
-      z_br_edges_inv[edge.id()] = {inst, true};
+      auto edge{dec_ctx.z3_ctx.bool_const(name.c_str())};
+      dec_ctx.z3_br_edges[{inst, cond}] = dec_ctx.z3_exprs.size();
+      dec_ctx.z3_exprs.push_back(edge);
+      dec_ctx.z3_br_edges_inv[edge.id()] = {inst, true};
     } else {
-      auto edge{!(GetOrCreateEdgeForBranch(inst, true))};
-      z_br_edges[{inst, cond}] = z_exprs.size();
-      z_exprs.push_back(edge);
+      auto edge{!(ToExpr(GetOrCreateEdgeForBranch(inst, true)))};
+      dec_ctx.z3_br_edges[{inst, cond}] = dec_ctx.z3_exprs.size();
+      dec_ctx.z3_exprs.push_back(edge);
     }
   }
 
-  return z_exprs[z_br_edges[{inst, cond}]];
+  return dec_ctx.z3_br_edges[{inst, cond}];
 }
 
-z3::expr GenerateAST::GetOrCreateEdgeForSwitch(llvm::SwitchInst *inst,
+unsigned GenerateAST::GetOrCreateVarForSwitch(llvm::SwitchInst *inst) {
+  // To aide simplification, switch instructions actually produce numerical
+  // variables instead of boolean ones, but are always compared against a
+  // constant value.
+  if (dec_ctx.z3_sw_vars.find(inst) == dec_ctx.z3_sw_vars.end()) {
```

Comment:
Let's do an early return here.

Reply:
Would
```
  if (dec_ctx.z3_sw_vars.find(inst) == dec_ctx.z3_sw_vars.end()) {
    auto name{GetName(inst)};
    auto var{dec_ctx.z3_ctx.int_const(name.c_str())};
    dec_ctx.z3_sw_vars[inst] = dec_ctx.z3_exprs.size();
    dec_ctx.z3_exprs.push_back(var);
    dec_ctx.z3_sw_vars_inv[var.id()] = inst;
  }
  return dec_ctx.z3_sw_vars[inst];
```
be ok?

Reply:
I usually prefer it the other way around cause usually the big block of code is gonna contain long lines that benefit from one less indent. So, `if (dec_ctx.z3_sw_vars.find(inst) != dec_ctx.z3_sw_vars.end()) { return ...; } ...`